### PR TITLE
[student-libraries] Added CloudFormation template for the new student libraries bucket

### PR DESCRIPTION
**Pull request by jmkulwik**
We tested this template and created a sample libraries bucket and stack. After this PR is approved, I will delete those test artifacts and create the final version.

The stack currently lives here: https://console.aws.amazon.com/cloudformation/home?region=us-east-1#/stacks/changesets?filteringText=&filteringStatus=active&viewNested=true&hideStacks=false&stackId=arn%3Aaws%3Acloudformation%3Aus-east-1%3A475661607190%3Astack%2Ftest-libraries-bucket%2F236efd30-da61-11e9-8642-12f81177eb92 

The bucket currently lives here: https://s3.console.aws.amazon.com/s3/buckets/cdo-v3-libraries/?region=us-east-1&tab=overview

Note that the bucket currently has "Object-level logging" set to enabled. Although we don't think this is necessary for this bucket, it appears to be a default setting, and we left it enabled.

Additional permissions should be set here prior to this PR being merged: https://github.com/code-dot-org/code-dot-org/blob/ca6bdeb6c5c6571ac4b4a03eba53bcd37372747c/aws/cloudformation/cloud_formation_stack.yml.erb#L139

### CODE REVIEW

**Comment by jmkulwik** [comment]
@wjordan, Suresh and I weren't sure. What would you recommend is the best way to make changes to the cloud formation stack? i.e. https://github.com/code-dot-org/code-dot-org/blob/ca6bdeb6c5c6571ac4b4a03eba53bcd37372747c/aws/cloudformation/cloud_formation_stack.yml.erb#L139

I'm not sure how we could test whether the changes worked due to the * permission.

**Comment by wjordan** [comment]
The S3 permissions on our application haven't yet been restricted to specific buckets, paths or operations - doing so will require tracking down _every_ S3 operation performed throughout our entire application codebase, and granting minimal permissions to only those buckets/paths/operations currently in use in each environment. Considering the complexity of our application, that work will probably be quite time-consuming and will introduce some production risk (of incorrect permissions causing breaking changes). The work hasn't been prioritized.

Until that broader work-item is done, my recommendation is to not make any changes to that line in the cloudformation stack- we can't make a change to only grant S3 permissions to only the path used by this new feature because it would break other parts of the application that implicitly depend on the current wild-card permission.

**Comment by sureshc** [comment]
@wjordan Could we explicitly grant permissions to each of the different managed environments to each of the different prefixes in this specific bucker in the IAM template without removing the explicit wildcard permissions?  I assume the wildcard would take precedence, but it would be nice to declare what we know about this new bucket and what permissions are needed for future work to make these permissions more granular.

**Comment by wjordan** [comment]
> it would be nice to declare what we know about this new bucket and what permissions are needed for future work to make these permissions more granular.

Sure, adding the permission on top of the wildcard wouldn't cause a problem- at this point it would just act as a kind of placeholder comment, there's no way we could test that the added permissions work correctly with the feature/application until we do the future work to remove the wildcard itself.

**Comment by sureshc** [comment]
Yup.  @jmkulwik noted that it can't be tested.  I think it's worth adding the permissions now for this bucket to move us towards the long term goal.

**Comment by jmkulwik** [comment]
> Yup. @ jmkulwik noted that it can't be tested. I think it's worth adding the permissions now for this bucket to move us towards the long term goal.

Sounds good! I'll add the permission. Please take a close look at it as there isn't a good way to live test it!

File: aws/cloudformation/cloud_formation_stack.yml.erb [hunk 1]
```diff
@@ -135,6 +135,17 @@ Resources:
...
+          - Effect: Allow
+            Action:
+              - 's3:DeleteObject'
+              - 's3:DeleteObjectVersion'
+              - 's3:GetObject'
+              - 's3:GetObjectVersion'
+              - 's3:PutObject'
+              - 's3:ReplicateObject'
+              - 's3:RestoreObject'
+            Resource: 'arn:aws:s3:::cdo-v3-libraries/libraries_<%=environment%>/*'
```
Comment (jmkulwik) [inline aws/cloudformation/cloud_formation_stack.yml.erb:148]: Docs for ^ are here: 
https://docs.aws.amazon.com/IAM/latest/UserGuide/list_amazons3.html#amazons3-actions-as-permissions
https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_examples_s3_rw-bucket.html

**Review by wjordan — APPROVED** [review]

**Pull request merged**